pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+                expr = RangedExpression( (repn0.to_expression(), repn1.to_expression(), repn2.to_expression()), strict=expr._strict )
+            else:
+                repn = generate_standard_repn(expr, quadratic=True, compute_values=compute_values)
+                expr = repn.to_expression()
+        except:     #pragma: no cover
+            #
+            # Generation of the standard repn will fail if the
+            # expression is uninitialized.  Hence, we default to
+            # using the non-standardized form.
+            #
+            # It might be smarter to raise errors for specific issues (e.g. uninitialized parameters).
+            # Let's see if we start seeing errors that are masked here.
+            #
+            pass
+    #
+    # Setup the symbol map
+    #
+    if labeler is not None:
+        if smap is None:
+            smap = SymbolMap()
+        smap.default_labeler = labeler
+    #
+    # Create and execute the visitor pattern
+    #
+    visitor = _ToStringVisitor(verbose, smap, compute_values)
+    return visitor.dfs_postorder_stack(expr)
+
+
+#-------------------------------------------------------
+#
+# Expression classes
+#
+#-------------------------------------------------------
+
+
+class ExpressionBase(NumericValue):
+    """
+    The base class for Pyomo expressions.
+
+    This class is used to define nodes in an expression
+    tree.
+
+    Args:
+        args (list or tuple): Children of this node.
+    """
+
+    # Previously, we used _args to define expression class arguments.
+    # Here, we use _args_ to force errors for code that was referencing this
+    # data.  There are now accessor methods, so in most cases users
+    # and developers should not directly access the _args_ data values.
+    __slots__ =  ('_args_',)
+    PRECEDENCE = 0
+
+    def __init__(self, args):
+        self._args_ = args
+
+    def nargs(self):
+        """
+        Returns the number of child nodes.
+
+        By default, Pyomo expressions represent binary operations
+        with two arguments.
+
+        Note:
+            This function does not simply compute the length of
+            :attr:`_args_` because some expression classes use
+            a subset of the :attr:`_args_` array.  Thus, it
+            is imperative that developers use this method!
+
+        Returns:
+            A nonnegative integer that is the number of child nodes.
+        """
+        return 2
+
+    def arg(self, i):
+        """
+        Return the i-th child node.
+
+        Args:
+            i (int): Nonnegative index of the child that is returned.
+
+        Returns:
+            The i-th child node.
+        """
+        if i >= self.nargs():
+            raise KeyError("Invalid index for expression argument: %d" % i)
+        if i < 0:
+            return self._args_[self.nargs()+i]
+        return self._args_[i]
+
+    @property
+    def args(self):
```

Comment:
Why is this defined this way in the base class?  Since Sum is the only thing where nargs and len(_args_) aren't the same, you could make this more efficient by making this implementation just iterate over _args_ and move the extra function calls and complexity to just Sum.